ta/gatekeeper/gatekeeper_ta.c

```diff
@@ -0,0 +1,783 @@
+				      expected_handle->user_id,
+				      expected_handle->flags,
+				      expected_handle->version,
+				      password, password_length);
+	if (res) {
+		EMSG("Failed to create password handle");
+		goto exit;
+	}
+
+	if (memcmp(password_handle.signature, expected_handle->signature,
+		   sizeof(expected_handle->signature)) == 0)
+		res = TEE_SUCCESS;
+	else
+		res = TEE_ERROR_SIGNATURE_INVALID;
+
+exit:
+	return res;
+}
+
+static TEE_Result TA_Enroll(TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_SUCCESS;
+
+	/*
+	 * Enroll request layout
+	 * +--------------------------------+---------------------------------+
+	 * | Name                           | Number of bytes                 |
+	 * +--------------------------------+---------------------------------+
+	 * | uid                            | 4                               |
+	 * | desired_password_length        | 4                               |
+	 * | desired_password               | #desired_password_length        |
+	 * | current_password_length        | 4                               |
+	 * | current_password               | #current_password_length        |
+	 * | current_password_handle_length | 4                               |
+	 * | current_password_handle        | #current_password_handle_length |
+	 * +--------------------------------+---------------------------------+
+	 */
+	uint32_t uid;
+	uint32_t desired_password_length;
+	const uint8_t *desired_password;
+	uint32_t current_password_length;
+	const uint8_t *current_password;
+	uint32_t current_password_handle_length;
+	const uint8_t *current_password_handle;
+
+	const uint8_t *request = (const uint8_t *)params[0].memref.buffer;
+	const uint8_t *i_req = request;
+	struct failure_record record;
+
+	/*
+	 * Enroll response layout
+	 * +--------------------------------+---------------------------------+
+	 * | Name                           | Number of bytes                 |
+	 * +--------------------------------+---------------------------------+
+	 * | error                          | 4                               |
+	 * +--------------------------------+---------------------------------+
+	 * | timeout                        | 4                               |
+	 * +------------------------------ OR --------------------------------+
+	 * | password_handle_length         | 4                               |
+	 * | password_handle                | #password_handle_length         |
+	 * +--------------------------------+---------------------------------+
+	 */
+	uint32_t error = GK_ERROR_NONE;
+	uint32_t timeout = 0;
+	struct password_handle password_handle;
+
+	uint8_t *response = params[1].memref.buffer;
+	uint8_t *i_resp = response;
+
+	const uint32_t max_response_size = sizeof(uint32_t) +
+		sizeof(uint32_t) +
+		sizeof(struct password_handle);
+
+	secure_id_t user_id = 0;
+	uint64_t flags = 0;
+	salt_t salt;
+
+	deserialize_int(&i_req, &uid);
```

Comment:
These deserialize should always check the offset in buffer against `params[0].memref.size` as mentioned in the comment on `deserialize_*()`.